File: linkmint-backend/eventbus-python/src/linkmint_eventbus/consumer.py
"""Async Kafka consumer-group member (aiokafka) with commit-after-handle (at-least-once)."""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any

from aiokafka import AIOKafkaConsumer
from opentelemetry import context, trace
from opentelemetry.context import Context
from opentelemetry.propagate import extract
from opentelemetry.trace import SpanKind

from .envelope import Envelope
from .logging import get_logger
from .metrics import BUS_MESSAGES_CONSUMED

# A handler receives the logical event name and the decoded payload, mirroring the services'
# existing typed ``handle(name, payload)`` chokepoints.
HandleFunc = Callable[[str, dict[str, Any]], Awaitable[Any]]


class KafkaConsumer:
    """Decodes envelopes and dispatches them to a handler, committing offsets only after a clean
    handle. Within a partition it commits the clean prefix and stops at the first handle error (the
    failed message redelivers); an undecodable message is logged and skipped+committed (poison
    safe). Duplicates are possible (retry/rebalance), so handlers MUST be idempotent.
    """

    def __init__(
        self,
        brokers: list[str],
        topics: list[str],
        group_id: str,
        client_id: str = "linkmint",
        log: Any | None = None,
    ) -> None:
        self._consumer = AIOKafkaConsumer(
            *topics,
            bootstrap_servers=brokers,
            group_id=group_id,
            client_id=client_id,
            enable_auto_commit=False,
            # A new group reads from the earliest record so a newly-added consumer never silently
            # skips events produced before it joined; once it commits it resumes from there.
            auto_offset_reset="earliest",
        )
        self._log = log or get_logger("eventbus.consumer")

    async def run(self, handle: HandleFunc) -> None:
        """Poll and dispatch until the task is cancelled; then stop the consumer cleanly."""
        await self._consumer.start()
        try:
            while True:
                batches = await self._consumer.getmany(timeout_ms=1000, max_records=256)
                for tp, msgs in batches.items():
                    await self._process_partition(tp, msgs, handle)
        finally:
            await self._consumer.stop()

    async def _process_partition(self, tp: Any, msgs: list[Any], handle: HandleFunc) -> None:
        commit_offset: int | None = None
        for msg in msgs:
            try:
                env = Envelope.from_bytes(msg.value)
            except Exception:  # noqa: BLE001 — undecodable bytes can never succeed
                self._log.warning("eventbus_decode_failed", topic=msg.topic, offset=msg.offset)
                commit_offset = msg.offset + 1  # poison-safe: skip + commit
                continue
            # Continue the producer's trace under a consume span; the handler runs within it.
            span = trace.get_tracer("eventbus").start_span(
                f"consume {env.name}",
                context=_extract_trace(getattr(msg, "headers", None)),
                kind=SpanKind.CONSUMER,
            )
            token = context.attach(trace.set_span_in_context(span))
            failed = False
            try:
                await handle(env.name, env.payload)
            except Exception as exc:  # noqa: BLE001 — any handler error → do not commit → redeliver
                failed = True
                BUS_MESSAGES_CONSUMED.labels(env.name, "error").inc()
                self._log.warning(
                    "eventbus_handle_failed",
                    name=env.name,
                    key=env.key,
                    offset=msg.offset,
                    error=str(exc),
                )
            else:
                BUS_MESSAGES_CONSUMED.labels(env.name, "ok").inc()
            finally:
                span.end()
                context.detach(token)
            if failed:
                break  # stop this partition; the failed message + the rest redeliver
            commit_offset = msg.offset + 1
        if commit_offset is not None:
            await self._consumer.commit({tp: commit_offset})


def _extract_trace(headers: Any) -> Context:
    """Rebuild the W3C trace context from a Kafka message's headers (list of (str, bytes) pairs)."""
    carrier: dict[str, str] = {}
    for item in headers or []:
        k, v = item
        carrier[k] = v.decode("utf-8") if isinstance(v, (bytes, bytearray)) else str(v)
    return extract(carrier)

File: linkmint-backend/eventbus-go/consumer.go
package eventbus

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"

	"github.com/twmb/franz-go/pkg/kgo"
)

// HandleFunc processes one event: its logical name and canonical payload bytes. Returning an error
// means "not handled" — the offset is NOT committed and the event is redelivered (at-least-once), so
// handlers MUST be idempotent. A nil return makes the event committable.
type HandleFunc func(ctx context.Context, name string, payload json.RawMessage) error

// Consumer is a Kafka consumer-group member that decodes envelopes and dispatches them to a
// HandleFunc, committing offsets only after a successful handle.
type Consumer struct {
	client *kgo.Client
	log    *slog.Logger
}

// NewConsumer joins the configured consumer group, subscribed to the given domain topics.
func NewConsumer(cfg Config, topics []string, log *slog.Logger) (*Consumer, error) {
	if log == nil {
		log = slog.Default()
	}
	if len(cfg.Brokers) == 0 {
		return nil, fmt.Errorf("eventbus: no kafka brokers configured")
	}
	if cfg.GroupID == "" {
		return nil, fmt.Errorf("eventbus: consumer group id is required")
	}
	if len(topics) == 0 {
		return nil, fmt.Errorf("eventbus: at least one topic is required")
	}
	cl, err := kgo.NewClient(
		kgo.SeedBrokers(cfg.Brokers...),
		kgo.ClientID(cfg.ClientID),
		kgo.ConsumerGroup(cfg.GroupID),
		kgo.ConsumeTopics(topics...),
		kgo.DisableAutoCommit(),
		// A group with no committed offset starts at the earliest record, so a newly-added consumer
		// never silently skips events produced before it joined (at-least-once; idempotent handlers
		// absorb any replay). Once the group commits, it resumes from its committed offset.
		kgo.ConsumeResetOffset(kgo.NewOffset().AtStart()),
	)
	if err != nil {
		return nil, fmt.Errorf("eventbus: new consumer: %w", err)
	}
	return &Consumer{client: cl, log: log}, nil
}

// Run polls and dispatches until ctx is cancelled, then closes the client. Within a partition it
// processes records in offset order and commits the clean prefix; the first handle error stops that
// partition for the round (the failed record + everything after it redeliver). A record whose bytes
// fail to decode is logged and skipped+committed (poison-safe) — it can never succeed and would
// otherwise block the partition forever.
func (c *Consumer) Run(ctx context.Context, handle HandleFunc) error {
	defer c.client.Close()
	for {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		fetches := c.client.PollRecords(ctx, 256)
		if fetches.IsClientClosed() {
			return nil
		}
		fetches.EachError(func(t string, p int32, err error) {
			if !errors.Is(err, context.Canceled) {
				c.log.Warn("eventbus_fetch_error", "topic", t, "partition", p, "err", err.Error())
			}
		})
		var commit []*kgo.Record
		fetches.EachPartition(func(ftp kgo.FetchTopicPartition) {
			for _, rec := range ftp.Records {
				env, err := UnmarshalEnvelope(rec.Value)
				if err != nil {
					c.log.Warn("eventbus_decode_failed", "topic", rec.Topic, "offset", rec.Offset, "err", err.Error())
					commit = append(commit, rec) // poison-safe: skip + commit
					continue
				}
				// Continue the producer's trace under a consume span; the handler runs within it.
				recCtx, span := startConsumeSpan(ctx, env.Name, rec.Headers)
				herr := handle(recCtx, env.Name, env.Payload)
				span.End()
				if herr != nil {
					c.log.Warn("eventbus_handle_failed", "name", env.Name, "key", env.Key, "offset", rec.Offset, "err", herr.Error())
					return // stop this partition; do not commit the failed record → it redelivers
				}
				commit = append(commit, rec)
			}
		})
		if len(commit) > 0 {
			if err := c.client.CommitRecords(ctx, commit...); err != nil && !errors.Is(err, context.Canceled) {
				c.log.Warn("eventbus_commit_failed", "err", err.Error())
			}
		}
	}
}
